### PR TITLE
Feature TLS Integration tests

### PR DESCRIPTION
**Pull request by paulomach**
# Issue

Implementation of TLS must be tested in conjunction with the TLS Certificates operator charm.
Task reference is: [DPE-628](https://warthogs.atlassian.net/browse/DPE-628)

This must be merged after #48 and this [MP](https://code.launchpad.net/~paulomachado/snap-mysql-shell/+git/snap-mysql-shell/+merge/431297) - tests will fail before that

# Solution

Tests were implemented based on mongodb-operator tests. (Shout out to @MiaAltieri )

# Context

Test connections before enabling TLS, after enabling TLS (by relating to TLS charm), common operations with TLS enabled (e.g. private key rotation) and test connections after breaking TLS charm relation.

# Testing

Implemented in `tests/integration/test_tls.py`

# Release Notes

* `tests/integration/test_tls.py` - The tests
* `src/charm.py` - refactor rolling restart routine
* `tests/integration/helpers.py` - new helpers and extended features

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -261,7 +262,7 @@ def _on_database_storage_detaching(self, _) -> None:
     def _on_update_status(self, _) -> None:
         """Handle update status.
 
-        Takes care of workload health checks.
+        Takes care of workload health checks and.
```
Comment (marceloneppel) [inline src/charm.py:265]: Is there something missing here?

File: tests/integration/helpers.py [hunk 2]
```diff
@@ -770,3 +772,49 @@ async def write_random_chars_to_test_table(ops_test: OpsTest, primary_unit: Unit
...
+    # Filter the data based on the relation name.
+    relation_data = [
+        v for v in data[unit_name]["relation-info"] if v["endpoint"] == "certificates"
+    ]
+    if len(relation_data) == 0:
+        return ""
+    return json.loads(relation_data[0]["application-data"]["certificates"])[0].get("ca")
+
+
+async def unit_file_md5(ops_test: OpsTest, unit_name: str, file_path: str) -> str:
```
Comment (marceloneppel) [inline tests/integration/helpers.py:803]: Nice!

File: tests/integration/test_tls.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+    for unit in all_units:
+        new_cert_md5 = await unit_file_md5(
+            ops_test, unit.name, f"/var/lib/mysql/{TLS_SSL_CERT_FILE}"
+        )
+        new_mysql_pid = await get_process_pid(ops_test, unit.name, "mysqld")
+
+        assert (
+            new_cert_md5 != original_tls[unit.name]["cert"]
+        ), f"cert for {unit.name} was not updated."
+        assert new_mysql_pid > original_tls[unit.name]["mysql_pid"], "❌ mysqld was not restarted"
```
Comment (marceloneppel) [inline tests/integration/test_tls.py:181]: Very thorough!

**Review by marceloneppel — APPROVED** [review]
Very well tested, Paulo! Do you plan to add the job to `.github/workflows/ci.yaml` as you did on https://github.com/canonical/mysql-k8s-operator/pull/104?

File: src/charm.py (hunk 1 above)
Comment (paulomach) [reply src/charm.py:265]: yes (:facepalm:)

**Comment by paulomach** [comment]
> Very well tested, Paulo! Do you plan to add the job to `.github/workflows/ci.yaml` as you did on [canonical/mysql-k8s-operator#104](https://github.com/canonical/mysql-k8s-operator/pull/104)?

Of course I forgot it!

**Review by marceloneppel — APPROVED** [review]

File: lib/charms/mysql/v0/mysql.py [hunk 4]
```diff
@@ -683,14 +683,15 @@ def is_instance_in_cluster(self, unit_label: str) -> bool:
             )
             return False
 
-    def get_cluster_status(self) -> dict:
```
Comment (shayancanonical) [inline lib/charms/mysql/v0/mysql.py:686]: should we still update the lib patch?

File: src/charm.py [hunk 5]
```diff
@@ -509,12 +524,25 @@ def _get_primary_address_from_peers(self) -> str:
                 return self.peers.data[unit]["instance-hostname"]
 
     def _restart(self, _) -> None:
+        """Restart server rolling ops callback function.
+
+        Hold execution until server is back in the cluster.
+        Used exclusively for rolling restarts.
+        """
+        logger.debug("Restarting mysqld daemon")
         if service_restart(SERVICE_NAME):
```
Comment (shayancanonical) [inline src/charm.py:533]: what should we do if the service restart is not successful?

**Review by shayancanonical — APPROVED** [review]
LGTM! Just have a couple of minor comments

File: src/charm.py (hunk 5 above)
Comment (paulomach) [reply src/charm.py:533]: Well spotted. Definitely block the unit ( 7a72c73 )

**Pull request merged**